data-model/src/main/java/org/hypertrace/core/datamodel/shared/StructuredTraceGraph.java

```diff
@@ -113,59 +127,69 @@ private void buildParentChildRelationship(StructuredTrace trace) {
       return;
     }
 
-    for (Entity entity : entities) {
-      entityMap.put(entity.getEntityId(), entity);
-    }
-
     List<Edge> entityEdges = trace.getEntityEdgeList();
     if (entityEdges != null) {
       for (Edge entityEdge : trace.getEntityEdgeList()) {
         Integer sourceIndex = entityEdge.getSrcIndex();
         Integer targetIndex = entityEdge.getTgtIndex();
         Entity parentEntity = entities.get(sourceIndex);
         Entity childEntity = entities.get(targetIndex);
-        parentToChildrenEntities.putIfAbsent(parentEntity, new ArrayList<>());
-        parentToChildrenEntities.get(parentEntity).add(childEntity);
-        childToParentEntities.putIfAbsent(childEntity, new ArrayList<>());
-        childToParentEntities.get(childEntity).add(parentEntity);
+        parentToChildrenEntities.computeIfAbsent(parentEntity.getEntityId(), k -> new ArrayList<>())
+            .add(childEntity);
+        childToParentEntities.computeIfAbsent(childEntity.getEntityId(), k -> new ArrayList<>())
+            .add(parentEntity);
       }
     }
   }
 
   private void buildRootEvents(StructuredTrace trace) {
```

Comment:
Is it possible to club this method with `buildParentChildRelationship`?
In `buildParentChildRelationship` we are iterating over the list of edges, and we have reference to all the child events, we could just collect all the child events here. Once we are done building the parent to child relation the removal of entities can be done from the rootEventIds.

Reply:
We can do that but I guess it won't make much difference in terms of efficiency and feel like this way(having separate methods doing separate things) is cleaner.

Reply:
Makes sense.